cli/emit.rs

```diff
@@ -378,57 +298,48 @@ pub struct CheckOptions {
   pub type_check_mode: TypeCheckMode,
   /// Set the debug flag on the TypeScript type checker.
   pub debug: bool,
-  /// If true, any files emitted will be cached, even if there are diagnostics
-  /// produced. If false, if there are diagnostics, caching emitted files will
-  /// be skipped.
-  pub emit_with_diagnostics: bool,
   /// The module specifier to the configuration file, passed to tsc so that
   /// configuration related diagnostics are properly formed.
   pub maybe_config_specifier: Option<ModuleSpecifier>,
   /// The derived tsconfig that should be used when checking.
   pub ts_config: TsConfig,
   /// If true, `Check <specifier>` will be written to stdout for each root.
   pub log_checks: bool,
-  /// If true, valid existing emits and `.tsbuildinfo` files will be ignored.
+  /// If true, valid `.tsbuildinfo` files will be ignored.
   pub reload: bool,
-  pub reload_exclusions: HashSet<ModuleSpecifier>,
 }
 
-/// The result of a check or emit of a module graph. Note that the actual
-/// emitted sources are stored in the cache and are not returned in the result.
+/// The result of a check of a module graph.
 #[derive(Debug, Default)]
-pub struct CheckEmitResult {
+pub struct CheckResult {
   pub diagnostics: Diagnostics,
   pub stats: Stats,
 }
 
-/// Given a set of roots and graph data, type check the module graph and
-/// optionally emit modules, updating the cache as appropriate. Emitting is
-/// determined by the `ts_config` supplied in the options, and if emitting, the
-/// files are stored in the cache.
+/// Given a set of roots and graph data, type check the module graph.
 ///
 /// It is expected that it is determined if a check and/or emit is validated
 /// before the function is called.
-pub fn check_and_maybe_emit(
+pub fn check(
   roots: &[(ModuleSpecifier, ModuleKind)],
   graph_data: Arc<RwLock<GraphData>>,
-  cache: &dyn EmitCache,
+  cache: &TypeCheckCache,
   options: CheckOptions,
-) -> Result<CheckEmitResult, AnyError> {
+) -> Result<CheckResult, AnyError> {
   let check_js = options.ts_config.get_check_js();
   let segment_graph_data = {
     let graph_data = graph_data.read();
     graph_data.graph_segment(roots).unwrap()
   };
-  if valid_emit(
-    &segment_graph_data,
-    cache,
-    &options.ts_config,
-    options.reload,
-    &options.reload_exclusions,
-  ) {
+  let check_hash = match get_check_hash(&segment_graph_data, &options) {
+    CheckHashResult::NoFiles => return Ok(Default::default()),
+    CheckHashResult::Hash(hash) => hash,
+  };
+  // do not type check if we know this is type checked
+  if cache.has_check_hash(check_hash) {
```

Comment:
This should respect `--reload`
```suggestion
  if !(options.reload
    && !roots
      .iter()
      .all(|r| options.reload_exclusions.contains(&r.0)))
    && cache.has_check_hash(check_hash)
  {
```

Reply:
Reload exclusions seems like too much complexity and I'm not sure what it does. Let's just always type check if someone provides `--reload`.